File: backend/src/main/java/com/bbc/app/service/CustomerService.java
package com.bbc.app.service;

import com.bbc.app.dto.response.CustomersResponse;
import com.bbc.app.dto.response.MessageResponse;
import com.bbc.app.dto.response.StatisticsResponse;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;

@Service
public interface CustomerService {

    ResponseEntity<StatisticsResponse> getTotalCustomersCount();

    // Retrieve all customers
    ResponseEntity<CustomersResponse> getAllCustomers(int page, int size);

    ResponseEntity<CustomersResponse> getCustomersByMeterNo(String meterNo, int page, int size);

    // Create a new customer
    ResponseEntity<MessageResponse> createCustomer(String name, String email, String phone, String address);

    // Update an existing customer
    ResponseEntity<MessageResponse> updateCustomer(String meterno, String name, String email, String phone, String address);

    // Delete a customer
    ResponseEntity<MessageResponse> deleteCustomer(String meterno);

    //check if user has any unpaid invoices pending
    public boolean hasUnpaidInvoices(String meterNo);

    // Bulk upload customers
    ResponseEntity<MessageResponse> bulkUploadCustomer(MultipartFile dataFile) throws IOException;
}

File: frontend/src/app/shared/services/customer.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CreateCustomerRequest, CustomersResponse, MessageResponse, SingleCustomerResponse, UpdateCustomerRequest } from '../../types/customer';

@Injectable({
  providedIn: 'root',
})
export class CustomerService {
  private apiUrl = 'http://localhost:8080/api/customers'; // Backend URL

  constructor(private http: HttpClient) {}

  // Get all customers with pagination
  getAllCustomers(page: number = 0, size: number = 5): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}?page=${page}&size=${size}`);
  }

  
  // Create a customer
  createCustomer(request: CreateCustomerRequest): Observable<MessageResponse> {
    return this.http.post<MessageResponse>(this.apiUrl, request);
  }

  // Update an existing customer
  updateCustomer(meterNo: string, request: UpdateCustomerRequest): Observable<MessageResponse> {
    return this.http.put<MessageResponse>(`${this.apiUrl}/${meterNo}`, request);
  }

  // Delete a customer by meter number
  deleteCustomer(meterNo: string): Observable<MessageResponse> {
    return this.http.delete<MessageResponse>(`${this.apiUrl}/${meterNo}`);
  }

  // Bulk upload customers from a file
  bulkUploadCustomer(file: File): Observable<MessageResponse> {
    const formData: FormData = new FormData();
    formData.append('file', file, file.name);
    
    return this.http.post<MessageResponse>(`${this.apiUrl}/upload`, formData);
  }

  // Get customers by meter no with pagination
  getCustomerByMeterNo(meterNo: string): Observable<SingleCustomerResponse> {
    return this.http.get<SingleCustomerResponse>(`${this.apiUrl}/${meterNo}`);
  }

  
}
